pkg/route/apiserver/registry/route/strategy.go

```
@@ -61,6 +61,10 @@ func (s routeStrategy) PrepareForCreate(ctx context.Context, obj runtime.Object)
 	route := obj.(*routeapi.Route)
 	route.Status = routeapi.RouteStatus{}
 	stripEmptyDestinationCACertificate(route)
+
+	if !s.allowExternalCertificates && route.Spec.TLS != nil {
+		route.Spec.TLS.ExternalCertificate.Name = ""
```

why not to return an error ?

---

> why not to return an error ?

Because the way kube APIs work, if newer manifests are used against older servers, the unknown fields are stripped (since before v1).  We could change the semantic and there are some valid reasons to do so (the object you wanted wasn't created), but it is lossy since the skew is still only at the point where the type can be deserialized.  This is backwards compatible with 4.13 behavior for instance.

---

I understand that the old kas will drop the field during decoding. 
I also understand that we are introducing a new field to the existing type, and we want to have it hidden behind a feature gate (FG).

What I don't understand is:
- Why don't we stop accepting data into this field until the FG is enabled?
- Why are we dropping data on an update? (Such a solution seems counterintuitive, and I think it's not what users expect)
 
In my opinion, what we should do is:
- Don't accept (return an error) data into this field unless the FG is enabled.
- If some objects with the new field set already exist in etcd, we should allow reading and writing them (update) without dropping the data (after the FG was on).

I believe this approach would be acceptable because, after the first release, all servers will have this new field (allowing roundtrip etcd data), but during validation, we won't allow data into this field until the FG feature is enabled (which occurs after a cluster upgrade)

---

I discussed this point with David in Slack. 
Techpreview feature gates are one-way switch.
We also don't allow turning it back off and we don't allow upgrades.
It is also not possible to drop fields on read because we don't have an injection point for that.